File: android/app/src/main/java/com/aikeyboard/PredictionService.kt
package com.aikeyboard

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class PredictionService(private val apiService: ApiService) {
    
    suspend fun getPredictions(text: String, language: String = "en"): List<String> {
        return try {
            val response = apiService.getPredictions(text, language)
            response.map { it.text }
        } catch (e: Exception) {
            // Fallback to local predictions
            getLocalPredictions(text, language)
        }
    }
    
    private fun getLocalPredictions(text: String, language: String): List<String> {
        val commonPhrases = mapOf(
            "en" to listOf("hello", "thank you", "please", "how are you", "I think"),
            "ru" to listOf("привет", "спасибо", "пожалуйста", "как дела", "я думаю"),
            "es" to listOf("hola", "gracias", "por favor", "cómo estás", "creo que"),
            "de" to listOf("hallo", "danke", "bitte", "wie geht es", "ich denke")
        )
        
        val phrases = commonPhrases[language] ?: commonPhrases["en"]!!
        
        return phrases.filter { phrase ->
            phrase.lowercase().startsWith(text.lowercase()) || text.length < 2
        }.take(5)
    }
    
    suspend fun autocorrect(text: String, language: String = "en"): AutocorrectResult {
        return try {
            apiService.correctText(mapOf("text" to text, "language" to language))
        } catch (e: Exception) {
            AutocorrectResult(text, text, 0)
        }
    }
    
    suspend fun translateText(text: String, sourceLang: String, targetLang: String): TranslationResult {
        return try {
            apiService.translateText(mapOf(
                "text" to text,
                "sourceLang" to sourceLang,
                "targetLang" to targetLang
            ))
        } catch (e: Exception) {
            TranslationResult(text, text, sourceLang, targetLang)
        }
    }
}

File: backend/src/services/predictionService.js
class PredictionService {
  static async getPredictions(text, language = 'en') {
    const predictions = [];
    const lastWords = text.split(' ').slice(-3).join(' ');
    
    const commonPatterns = {
      'en': ['hello', 'thank you', 'please', 'sorry', 'how are you', 'I think', 'I want', 'can you', 'would you'],
      'ru': ['привет', 'спасибо', 'пожалуйста', 'извините', 'как дела', 'я думаю', 'я хочу', 'можете', 'хотите'],
      'es': ['hola', 'gracias', 'por favor', 'lo siento', 'cómo estás', 'creo que', 'quiero', 'puedes', 'gustaría'],
      'de': ['hallo', 'danke', 'bitte', 'entschuldigung', 'wie geht es', 'ich denke', 'ich möchte', 'können Sie', 'würden Sie']
    };
    
    const phrases = commonPatterns[language] || commonPatterns['en'];
    
    phrases.forEach(phrase => {
      if (phrase.toLowerCase().startsWith(lastWords.toLowerCase()) || lastWords.length < 2) {
        predictions.push({
          text: phrase,
          confidence: Math.random() * 0.3 + 0.7
        });
      }
    });
    
    predictions.sort((a, b) => b.confidence - a.confidence);
    return predictions.slice(0, 5);
  }
  
  static async getNextWordPrediction(text, language = 'en') {
    const words = text.split(' ');
    const lastWord = words[words.length - 1];
    
    const nextWords = await this.getPredictions(lastWord, language);
    return nextWords.map(p => p.text);
  }
  
  static async getSentenceCompletion(text, language = 'en') {
    const completions = await this.getPredictions(text, language);
    return completions.map(c => c.text);
  }
}

module.exports = PredictionService;